app/Http/Controllers/Chat/ChatController.php

```diff
@@ -123,65 +124,88 @@ public function __construct()
      */
     public function updates()
     {
-        $params = request()->all();
+        $request = request()->all();
+        $params = get_params($request, null, [
+            'channel_id:int',
+            'history_since:int',
+            'limit:int',
+            'since:int',
+        ], ['null_missing' => true]);
+
+        $includes = get_params($request, 'includes', [
+            'messages:bool',
+            'presence:bool',
+            'silences:bool',
+        ], ['null_missing' => true]);
 
-        if (!present($params['since'] ?? null)) {
+        if ($params['since'] === null) {
             abort(422);
         }
 
-        $presence = $this->presence();
-
         $since = $params['since'];
-        $limit = clamp(get_int($params['limit'] ?? null) ?? 50, 1, 50);
-
-        // this is used to filter out messages from restricted users/etc
-        $channelIds = array_map(function ($e) {
-            return $e['channel_id'];
-        }, $presence);
-
-        $messages = Message
-            ::with('sender')
-            ->whereIn('channel_id', $channelIds)
-            ->since($since)
-            ->limit($limit)
-            ->orderBy('message_id', 'DESC');
-
-        if (present($params['channel_id'] ?? null)) {
-            $messages->where('channel_id', get_int($params['channel_id']));
+        $limit = clamp($params['limit'] ?? 50, 1, 50);
+
+        $includeMessages = $includes['messages'] ?? true;
+        $includePresence = $includes['presence'] ?? true;
+        $includeSilences = $includes['silences'] ?? true;
+
+        $response = [];
+
+        // messages need presence
+        if ($includeMessages || $includePresence) {
+            $presence = $this->presence();
+        }
+
+        if ($includeMessages) {
+            $channelIds = array_pluck($presence, 'channel_id');
+
+            $messages = Message
+                ::with('sender')
+                ->whereIn('channel_id', $channelIds)
+                ->since($since)
+                ->limit($limit)
+                ->orderBy('message_id', 'DESC');
+
+            if ($params['channel_id'] !== null) {
+                $messages->where('channel_id', $params['channel_id']);
+            }
```

Comment:
looks like this can be combined with `$channelIds` above (`$channelIds = $params[...] ... ? [$params[...]] : array_...`)

and then the query & fetch can be done at once